### PR TITLE
Refactor parse tree to improve code clarity

### LINKED ISSUE #19: Refactor Loki parse tree adjustments as a separate stage of backend
There are a number of changes that this code makes to the Sigma parse tree prior to converting it into a LogQL query. These include replacing strings that include wildcards with regular expressions, and removing NOT conditions through the application of De Morgan's law. At the moment this is done as the query string is being generated - but this choice has lead to multiple bugs (i.e., #13, #18) where expectations about the structure of the tree are incorrect or are not being applied uniformly throughout. It seems likely that as this code changes over time, more errors will crop up because of this (and our code will further diverge from the TextQueryBackend class, potentially leading to improvements to pySigma being incompatible with our own).

I propose the addition of a separate stage within [convert_rule](https://github.com/grafana/pySigma-backend-loki/blob/main/sigma/backends/loki/loki.py#L550) that makes relevant adjustments the parse tree, which will also allow us to simplify (if not remove completely) some of the custom conversion code. This could potentially be done as a pipeline, or even before the pipeline stage, although at this point, I can't identify any types of pipeline that rely on the parse tree structure, so this may not be necessary.

### PR DESCRIPTION
**Pull request by kelnage**
The Sigma specification includes features that are not directly supported in LogQL (e.g., wildcard strings and NOT operators) but equivalent functionality can be achieved via other means. Previously, this backend made changes to the parse tree "on-the-fly", as they were encountered during the conversion process. This led to bugs, and a lack of certainty about edge cases being covered.

I have now refactored the parse tree modifications (changing strings that contain wildcards into regular expressions and eliminating NOTs via De Morgan's laws) into a single function, that is executed after the pipeline stage. I used a depth-first recursive function, as there will be no infinite depth parse trees (without infinite length inputs anyway 😜), which made it easier to pass the negation state down the relevant sub-trees.

There is one place where the negation state still needs to be actively handled during the rule processing: SigmaExpansion's (such as base64offset token conversion) are converted into OR'd (or AND'd, if negated) equalities as there isn't support for them in the backend - but that conversion doesn't happen until they are encountered. This was handled in [convert_condition_field_eq_expansion](https://github.com/grafana/pySigma-backend-loki/blob/a91ef07e49c61c943eedbff32607011cd0598cc7/sigma/backends/loki/loki.py#L843). 

There was also a significant amount of code that could be removed (~100 lines - 10%!) or simplified thanks to these changes. I imagine there may be other places where more of our custom code could be removed (using the default TextBackend implementation instead), but I wasn't able to immediately identify any more at this point.

Closes #19.

### CODE REVIEW

File: sigma/backends/loki/loki.py [hunk 1]
```diff
@@ -227,38 +194,70 @@ class LogQLBackend(TextQueryBackend):
...
-        str
-    ] = "?"  # Character used as single-character wildcard (replaced with .)
+    negated_line_filter_operator: ClassVar[Dict[str, str]] = {
+        "|=": "!=",
+        "!=": "|=",
+        "|~": "!~",
+        "!~": "|~",
+    }
+
+    currentTemplates: ClassVar[Union[bool, None]] = None
```
Comment (mostafa) [inline sigma/backends/loki/loki.py:205]: If we switch to Python 3.10, we can make all these type information more readable like below:
```suggestion
    currentTemplates: ClassVar[bool | None] = None
```

Comment (mostafa) [inline sigma/backends/loki/loki.py:205]: We don't have camel case in Python.
```suggestion
    current_templates: ClassVar[Union[bool, None]] = None
```

File: sigma/backends/loki/loki.py [hunk 2]
```diff
@@ -227,38 +194,70 @@ class LogQLBackend(TextQueryBackend):
...
+    # Leave this to be set by the below function
+    eq_token: ClassVar[str]
+    field_null_expression: ClassVar[str]
+    re_expression: ClassVar[str]
+    cidr_expression: ClassVar[str]
+    compare_op_expression: ClassVar[str]
+    compare_operators: ClassVar[Dict[SigmaCompareExpression.CompareOperators, str]]
+
+    @staticmethod
+    def setExpressionTemplates(negated: bool) -> None:
```
Comment (mostafa) [inline sigma/backends/loki/loki.py:215]:
```suggestion
    def set_expression_templates(negated: bool) -> None:
```

**Review by mostafa — CHANGES_REQUESTED** [review]
I like the cleanups and the code LGTM. However, you used camel case in Python, which needs to be fixed. I don't think you need another round of review, unless you change other parts of the code.

File: sigma/backends/loki/loki.py (hunk 1 above)
Comment (kelnage) [reply sigma/backends/loki/loki.py:205]: Yes, but pySigma and sigma-cli both support Python 3.8, where this syntax won't work.

Comment (mostafa) [reply sigma/backends/loki/loki.py]: Maybe we can ask them to upgrade.

**Review by mostafa — APPROVED** [review]

**Pull request merged**